### ISSUE TITLE
Abstract suggestion: Overall data model documentation

### ISSUE THREAD

**Issue by joepie91** [matrix-org/matrix-doc#1985]
Currently, the specification seems to be defined in a very piecemeal-y manner; each part of the API individually specifies what data it expects, what data it returns, and so on.

This makes it somewhat difficult to build well-architected-upfront homeservers and clients, as you'd have to read pretty much the entire specification to build up a comprehensive (mental) model of the different kinds of 'entities' in a Matrix system, and how they relate to each other.

I'd suggest adding some sort of high-level data model documentation, that describes in a relational manner what kind of different data types (users, messages, access tokens, etc.) exist in a Matrix system, and how they relate to other data types. This would still be an abstract model, leaving it up to applications how to implement and persist it exactly.

For example, the kind of model I'm thinking of might be something like this (represented as text):

----

### User

- username: String
- password: String
- devices: [Device]
- joined_rooms: [Room]

### Device

- user: User
- access_token: String
- display_name: String

### Room

- name: String
- users: [User]
- messages: [Message]

### Message

- body: String
- sender: User
- recipient: User | Room
- parents: [Message]

----

That's just a simple example, of course (and most likely wrong - I haven't actually read the entire spec yet!), but it hopefully illustrates what I mean. It doesn't necessarily need to be a text representation either, if eg. a visual representation could make it easier to understand.

This data model also wouldn't need to cover 'ephemeral' API data (like eg. the `inhibit_login` parameter in the `/register` endpoint), only the persistent data that a developer would need to model into their application architecture.

(This wouldn't have to be part of the specification either, it could act as a non-normative guide separate from it.)

**Comment by turt2live** [comment]
I believe this is what https://matrix.org/docs/spec/#architecture attempts to document - what exactly is missing there?

**Comment by joepie91** [comment]
Hmm. As far as I understand it, that only describes *some* parts of the data model; specifically those parts necessary for understanding how Matrix itself works. It's certainly useful information, but not comprehensive enough to define the *full* data model.

As a concrete (if a bit blunt) example: you couldn't fully model your relational database off the descriptions there.

Some specific examples of information deficiencies there:

- 'Access tokens' (for authentication purposes) are definitely a separate entity, relationally speaking, but are not mentioned in that architectural overview.
- The information there is *descriptive*, not structural; it doesn't contain a scannable listing of different 'fields' within an entity, for example, I would have to infer that from the text (and keep my own notes on that). That's helpful for understanding the protocol, but not for modelling one's own data layer. Hence my example in the original post, which is more of a linear, scannable list of data fields.
- It's unclear how different entities relate to each other, and in what quantities. Does an event have a sender and a recipient? Or can it have *multiple* recipients? Or even multiple senders? Are users joined directly to rooms, or do you need an intermediate 'presence' representation to encode their role *within* a room? And so on.
- What data fields does an 'event' have anyway? Can I represent any kind of message event with a single 'body' field? Or do I need to handle different data structures?

In short: I think the architecture section should remain structured as it currently is, as it (quite effectively!) communicates the overall protocol structure. The more boring data model listing I'm suggesting would be in *addition* to that, to aid in modeling one's *own* data persistence layer.
